google/cloud/cloud_testing/environment_variable_restore.cc

```diff
@@ -0,0 +1,35 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/cloud_testing/environment_variable_restore.h"
```

Comment:
`google/cloud/cloud_testing`? Can we just do `google/cloud/testing` or similar? I'm not sure what `cloud_testing` gets us over `testing`?

Reply:
Because this:

https://google.github.io/styleguide/cppguide.html#Namespace_Names

> Avoid nested namespaces that match well-known top-level namespaces. Collisions between namespace names can lead to surprising build breaks because of name lookup rules. In particular, do not create any nested std namespaces. Prefer unique project identifiers (websearch::index, websearch::index_util) over collision-prone names like websearch::util.

We have been using `google::cloud::foo::testing` and that conflicts with the top-level `::testing` namespace in Googletest.  I do not want to make the situation worse, so this library "starts right". Yes, we should go and fix the other libraries too.

Reply:
I'm not a big fan of `cloud_testing` because of the repetition of `cloud/cloud_testing`.

Can we use `test_util` or `testing_util` or `test_common` or similar to avoid the repetition? Or are these worse?

Also, should this be under `internal`, or is this considered to be a "public" API?

Just some thoughts, feel free to do what you feel is best, don't block on me.

Reply:
> Also, should this be under internal, or is this considered to be a "public" API?

They are not part of the public API.   We have not placed any of them inside `internal` so far, OTOH:
-  The `testing` directories are not part of the client library at all. They are not linked when you link the client library, they are not installed with the library.
- So with CMake you cannot link or compile against them if you use installed libraries.
- With Bazel you cannot link or compile against them either.
- If you are compiling with the library as a submodule you could use these things, but there is no way to stop a user in that case.

> I'm not a big fan of `cloud_testing` because of the repetition of `cloud/cloud_testing`.

Me neither.  `testing_util` is tempting... It is a bit dangerous because there may be a top-level namespace called `testing_util`, but it is not "well-known", so I guess it would be Okay.  I will make that change.